achen/integrate and implement open telemetry tracing

achen2289 opened the pull request:
Test plan -
- Set Open Telemetry tracer and ensured query state changes are tracked as well as query completes execution successfully.

```
== RELEASE NOTES ==

General Changes
* Add Open Telemetry tracer implementation to existing distributed tracing system
* Add new tracer type to system tracing configuration
* Update tracer interface and classes to contain tracer name and tracer token
* Add Open Telemetry dependencies
* TODO: add DataDog / general exporting capability for OTEL traces
```

mlyublena committed:
Move constant grouping keys above the group by

For example:

SELECT userid, 'constant val' AS extra, COUNT(1) as cnt from T GROUP BY 1, 2
into

SELECT userid,  'constant val' as extra, cnt from (SELECT userid, AS extra, COUNT(1) from T GROUP BY 1) s

vivek-bharathan committed:
Support Domain Expressions for GlueHiveMetastore

    This work adds support that converts expressions on partition columns
    into valid glue expresions to improve performance. The Domain for each
    partition column is converted, if possible, into a glue expression
    as  per

    https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-partitions.html#aws-glue-api-catalog-partitions-GetPartitions

    The implementation is best effort and may still return more partitions for the caller to filter out (HivePartitionManager)

Cherry-pick of https://github.com/trinodb/trino/commit/643bfd1b315af1fc3dd80081a01b3362eb823898

Co-authored by Sam Rash <sam.rash@gmail.com>

asjadsyed committed:
Support multi-coordinator while serving task info

pranjalssh committed:
Improve HBO overhead with LIMIT

pranjalssh committed:
Canonicalize domain predicate in TableScanNode

wanglinsong committed:
add init Jenkinsfile

Commit:
Fix incorrectly set wall time

Wall time should be set to elapsed time, not total scheduled time

Cherry pick of https://github.com/trinodb/trino/commit/7cc7b51a86d0935afddf199073cd942728671b30
Co-Authored-By: William Shen <willshen@users.noreply.github.com>

pranjalssh committed:
Order map entries during HBO Canonicalization

vivek-bharathan committed:
Add javadoc for Presto query submission protocol

jainavi17 committed:
Adding support of MAP_TOP_N_KEYS to Presto

fgwang7w committed:
Add javadoc for query dispatch management

wanglinsong committed:
update to build tar and docker image

pramodsatya committed:
Add window ranking function tests

feilong-liu committed:
Fix bug in test TestFilteredAggregations

The test uses QueryAssertions to compare the results between input SQl
queries and expected SQL queries. However, the QueryAssertions created
does not take the session properties specified in the test. Here I fix
it by passing the session properties when creating QueryAssertions.

feilong-liu committed:
Remove redundant distinct over group by

Remove distinct if the corresponding output is already distinct after a
group by operation.

tanjialiang committed:
Add property builder and populator

ctgraves committed:
Updated assert aggregation to include approximate equals

nastra committed:
Update Iceberg to 1.0.0

Mionsz committed:
[native] Add folly install job for script to work correctly

setup-centos.sh modified to include folly install job.

Signed-off-by: Milosz Linkiewicz <milosz.linkiewicz@intel.com>

Mionsz committed:
[native] Centos unified build and install process

Use cmake_install for unified build process of dependencies.

Signed-off-by: Milosz Linkiewicz <milosz.linkiewicz@intel.com>

pranjalssh committed:
Fix flaky HBO tests

spershin committed:
[native] Temporarily disable HttpTest.outstandingRequests.

cliandy commented on presto-main/pom.xml:
```diff
@@ -474,6 +474,41 @@
             <artifactId>ratis-common</artifactId>
             <optional>true</optional>
         </dependency>
+
```
do we need all of these imports?

achen2289 replied:
yes, in `OpenTelemetryBuilder` they are used

cliandy commented on presto-main/src/main/java/com/facebook/presto/tracing/OpenTelemetryBuilder.java:
```diff
@@ -0,0 +1,57 @@
+        OpenTelemetry openTelemetry;
+        if (tracing) {
+            openTelemetry = OpenTelemetrySdk.builder()
+                    .setTracerProvider(sdkTracerProvider)
```
in what case would we use instantiate this without having tracing enabled?

achen2289 replied:
had it there initially because of how i structured the tracer provider, but changed that. so there's no case to initialize it without tracing. i can remove it.

cliandy replied:
sounds good

cliandy replied:
sounds good, lets simplify this

cliandy commented on presto-main/src/main/java/com/facebook/presto/tracing/OpenTelemetryTracer.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    public void addPoint(String annotation)
+    {
+        startBlock(annotation, String.valueOf(System.nanoTime()));
```
I can understand why there's a point here, but wondering where this is used in the codebase? May seem to make more sense as just nesting of spans

cliandy replied:
the way this is called, annotation is being passed into `startBlock`'s `blockName` parameter, is this intentional?

achen2289 replied:
with the `SimpleTracer`, when it is initialized, a point is added which tracks that time. since there's no equivalent of a point for the otel tracer, i decided it could be a really short block. is that what you're asking? what would nesting of spans be for?

achen2289 replied:
yes that is intentional. it will essentially create a span whose name is the annotation (start tracing)

cliandy replied:
for example let's say that there's a point for query completed, one other way this can be designated could be a span from the beginning to the end of the query which would show that information. Wondering depending where this is actually used we can see if we can remove this concept.

cliandy replied:
annotation should probably be used as the same field, and wouldn't use the time as an annotation here;

let's create some way to have some unique placeholder here in the meantime, and maybe push here to see if we can remove this concept

achen2289 replied:
addPoint is only used when initializing the tracer and ending the trace, and the points created represent start trace time and end trace time. tracer currently looks like this.

trace start (point) ---- [waiting for prerequisites (span/block)] [dispatching] .... [finished] trace end

for this specific usage, i don't think we can remove add point, because i don't think trace start and end make sense to be represented as spans? like if the span is from trace start -> waiting for prerequisites, what's going on during that span? and for trace end, it's a point added directly after the finished span ends.

achen2289 replied:
and personally i don't think trace start and end are required (the spans matter more than when the start and end of trace are), so i don't think add point is necessary, but don't think they should be removed either?

achen2289 replied:
the annotation isn't really an annotation the current way it's used in the simple tracer. for example, in the simple tracer, for the block representing waiting for resources, `startBlock("waiting for resources", "")` is called. the call doesn't really take advantage of the annotation field. so i decided to use annotation for the block name field here. i personally don't think both the block name and annotation field are required.

achen2289 replied:
tldr; i get what you're saying about removing the add point functionality so i'll try to implement a placeholder for it

cliandy reviewed (changes requested):
some pieces here, but generally looking good!

spershin committed:
[native] Add two spill_memory_threshold session propertiess.

jainxrohit committed:
Use querytype as key in QueryExecutionFactory map

Using QueryType instead of statement class as key in the
QueryExecutionFactory.
We also added CONTROL as new QueryType to categorize statements
of transaction control and session control types.

jainxrohit committed:
Rename QueryPreparer to BuiltInQueryPreparer

jainxrohit committed:
Add QueryPreparer interface

Adding a QueryPreparer interface, all analyzers should implement this
interface to support query prepare functionality.

swapsmagic committed:
Fix weighted fair scheduling in disagg coordinator

In disagg coordinator, each coorinator is considering it's own running queries to decide which
resource group should run the next query in weighted fair scheduling. This lead to an issue when
with n coordinators each allowing single query to run for the resource group which has lower weight
and lead to n slots been used by the resource group even if it's weight will allow it to run lower than
n queries.

With the fix, we consider global running query count for the given resource group, which helps decide the
correct number of queries we should run for a lower weight resource group and allow more slots be given to
high priority resource group as per weighted fair scheduling.

miaoever committed:
Add custom plan translation in LocalExecutionPlanner

presto-release-bot committed:
Prepare for next development iteration - 0.279-SNAPSHOT

cliandy commented on presto-main/pom.xml:
```diff
@@ -474,6 +474,41 @@
+            <artifactId>opentelemetry-context</artifactId>
+        </dependency>
+
+<!--        <dependency>-->
```
remove comments at least from PR to prestodb/presto codebase if not needed

cliandy commented on presto-spi/src/main/java/com/facebook/presto/spi/tracing/TracerProvider.java:
```diff
@@ -19,5 +19,5 @@
      * @return A @Tracer that should be kept throughout the whole duration of tracing.
      */
-    Tracer getNewTracer();
+    Tracer getNewTracer(String tracerName, String traceToken);
```
is there a reason we need to pass in a `tracerName` if the instance of the tracer already determines that information?

cliandy replied:
also, are we getting a new tracer each time; is this why we are passing in the trace token?

achen2289 replied:
we have a singular tracer for all http requests. however, a new span is created for each request / query and they have a trace token attribute which is why we need to pass in `traceToken`

achen2289 replied:
we pass in `tracerName` because if we call this method from different modules, we want different tracers created. if we call getNewTracer with the same tracerName, it will return the same Open Telemetry tracer

Commit:
Create REMOVE_NULLS function

tanjialiang committed:
Refactor TestingResponse to extract logic to be customizable by users

tanjialiang committed:
Add more tests to TestPrestoSparkHttpWorkerClient.java

g1y committed:
Fix CI for next snapshot iteration PRs

The new release process for presto creates pull requests
that have an incremented snapshot which hasn't
been built by anything yet. There are tests that are building
test coverage but skipping presto server. This creates a problem
where they are requesting dependencies that aren't built yet.

We don't need test-coverage in these tests, so it's best if
we just skip building those modules.

g1y committed:
Fix CI for next snapshot iteration PRs

The new release process for presto creates pull requests
that have an incremented snapshot which hasn't
been built by anything yet. There are tests that are building
test coverage but skipping presto server. This creates a problem
where they are requesting dependencies that aren't built yet.

We don't need test-coverage in these tests, so it's best if
we just skip building those modules.

Commit:
[native] Allow overriding file system registration.

gggrace14 committed:
[native] Pass LocationHandle to HiveInsertTableHandle

Enable the ser/de of protocol::HiveInsertTableHandle.

Extract LocationHandle and HiveColumnHandles from
protocol::HiveOutputTableHandle and protocol::HiveInsertTableHandle,
and add them to Velox connector::HiveInsertTableHandle.
Pass them down to the plan TableWriteNode.

pgupta2 committed:
Add retryExecutionStrategy in client tags

Commit:
Expose Filter push down support

pranjalssh committed:
Use map sorting json config in FRC

tanjialiang committed:
Fetcher timeout should be at http level

Commit:
Add TypeParameters to SqlInvokedFunction

nizarhejazi committed:
Support STRPOS pushdown to Pinot; Fix missing configuration property pinot.query-options error

Commit:
Generalize existing SQL UDF functions to all types

swapsmagic committed:
Fix Disaggregated Coordinator Weighted Scheduling

Resource group manager was considering locally running queries count to decide on the group weight
and prioritze next running query or eligible group. With the fix, it considers all queries running in
the cluster for a given resource group to decide the scheduling weight.

feilong-liu committed:
Add optimizer information collector for optimizer logging

Add optimization information collector to session, so that it can be
used to collect information we want to log from optimizers.

feilong-liu committed:
Add logging for IterativeOptimizer

Pass information from iterative optimizer to the optimizer information collector.

Commit:
Put type translator back

Mionsz committed:
Build shared libs

Little improvement that include building shared libs if possible and exclude tests and examples build.

Mionsz committed:
Use cmake_install from velox

Use cmake_install from velox

Mionsz committed:
Source velox helper functions

Source Velox helper functions depending on file existence.

Mionsz committed:
Fix to sourcing

Fix to sourcing

Mionsz committed:
Change $0 to $BASH_SOURCE[0]

Change $0 to $BASH_SOURCE[0]

achen2289 committed:
implement open telemetry tracing plugin

support simple tracer in new open telemetry implementation

bump okhttp3 version to 4.5.0

use system val if client val not set

add support for context extraction from b3 propagation and overall interface

revert back to okhttp3 version 3.9.0 for this PR

implement W3C baggage header extraction

tracer handle implementation, remove context propagator from system property, general refactor

address nits and open telemetry baggage updates

tanjialiang committed:
Add native execution related properties to config classes

majetideepak committed:
[native] Fix test native failure

Commit:
Fix MIN(x, n) and MAX(x, n) on window functions

Commit:
Add ARRAY_MIN_BY, ARRAY_MAX_BY functions

jingmeih committed:
Add release notes for 0.278

spershin committed:
[native] Use Release configuration instead of Debug for native-test.

Commit:
[native] Use velox exception instead of std exception.
